### ISSUE TITLE
[FEATURE] ElasticsearchEmbeddingStore supports hybrid/mixed search 

### ISSUE THREAD

**Issue by fmx0717** [langchain4j/langchain4j#1666]
ElasticsearchEmbeddingStore supports mixed search of semantics and keywords

**Comment by zzutligang** [comment]
me too!

**Comment by fmx0717** [comment]
@langchain4j 
Will this feature be released in version 0.36? Is there an estimated time for the rollout plan?

**Comment by dadoonet** [comment]
@langchain4j I'm unsure about what we want to do here. Is it meant to be generic (for all embedding stores)? Like `ElasticsearchEmbeddingStore.search(EmbeddingSearchRequest searchRequest, IsTextMatch keywordSearchRequest)`...

What do you have in mind?

**Comment by langchain4j** [comment]
@dadoonet I am not completely sure, there are multiple options:

1. add `String query` field to the `EmbeddingSearchRequest`. If `EmbeddingStore` implementation supports hybrid search, then `query` can be used along with `queryEmbedding` to perform the search. I guess all specifics of full-text search (if any) can be configured when `ElasticsearchEmbeddingStore` is created. But then the name `EmbeddingStore` becomes too narrow and should be renamed into `SearchEngine` or something. Now is perharps not the time to go this way (we are working on releasing stable 1.0 at the end of year), I would keep it for 2025.

2. `ElasticsearchEmbeddingStore.search(EmbeddingSearchRequest searchRequest, IsTextMatch keywordSearchRequest)` or similar can be implemented for sure, but: 1) the same problem as above 2) there will be no way to plug this into `DefaultRetrievalAugmentor`, so users will have to implement RAG manually with quite a bit of boilerplate.

3. More generic `ContentRetriever` interface can be implemented (e.g. `ElasticsearchContentRetriever`, similar to `AzureAiSearchContentRetriever`). I assume most of the vector search logic from `ElasticsearchEmbeddingStore` can be reused, so it should be OK.

WDYT?

**Comment by langchain4j** [comment]
Hi @dadoonet, any thoughts regarding my previous comment? I am sure this is an important feature to include in LC4j 1.0

**Comment by dadoonet** [comment]
Hey.

I started a branch 12 days ago but did not find time yet to finish it. I'm resuming my work on this today.
Let see what will be the outcome.

FYI I went to the `ContentRetriever` route. Let see how this plays.

**Comment by langchain4j** [comment]
@dadoonet short update: I plan a release next week, so if you have something ready, please open a PR 🙏

**Comment by dadoonet** [comment]
No. It's in progress but this week has been like crazy. I might find some space middle of next week but I hardly expect that a merge will be possible before your release 😔

**Comment by langchain4j** [comment]
Hi @dadoonet, how are you doing?
I was thinking it would be great to demonstrate to LC4j users how to easily set up hybrid RAG locally.
If it could be a ready-to-use Elasticsearch integration, that would be awesome!
But it could also be an example/template (perhaps with a blog post) about how to implement it manually using the `ContentRetriever` interface backed by custom Elasticsearch querying logic. WDYT?

**Comment by langchain4j** [comment]
It would also be great to have it documented [here](https://docs.langchain4j.dev/tutorials/rag). It might even be a separate "Hybrid RAG" section. But please don't feel pressured, I’m just thinking out loud

**Comment by dliubarskyi** [comment]
Hi @dadoonet, any news? Thanks!

**Comment by l-trotta** [comment]
hey @dliubarskyi, is there still interest in implementing this? I can take over for @dadoonet

**Comment by dliubarskyi** [comment]
@l-trotta, yes, please go ahead! 🙏
